### PR TITLE
Improve defaults

### PR DESCRIPTION
**Pull request by ironcladlou**
Various fixes which make the image work out of the box and in various
hosted PaaS environments (e.g. Kubernetes).

* If the data volume doesn’t contain a `znc.conf`, populate the volume with a sane default configuration set. The default config has:
  * An admin user (username: admin, password: admin).
  * A new generated certificate.
  * A less restrictive port binding (8080).
* Expose the container port by default.
* Remove chown/chmod commands from entrypoint (these break most non-host mounted volume scenarios).
* Run as the ZNC user via USER directive rather than init process flags.

### CODE REVIEW

**Comment by ironcladlou** [comment]
These are things I found while trying to [deploy ZNC to OpenShift](https://github.com/ironcladlou/znc-openshift).

**Comment by ironcladlou** [comment]
The `EXPOSE` might not make sense for a user who decides to change the port binding.

**Comment by DarthGandalf** [comment]
What is special in what Kubernetes does?

> If the data volume doesn’t contain a znc.conf, populate the volume with a sane default configuration set.

Ok. But if user passes `--makeconf`, it shouldn't copy this default config, and immediately complain that config already exists.

> Expose the container port by default.

What difference does it make? Any port can be forwarded with `-p`, regardless of what ports are exposed.

> Remove chown/chmod commands from entrypoint (these break most non-host mounted volume scenarios).

Break how? What are these scenarios?
So what will happen if user is wrong?

> Run as the ZNC user via USER directive rather than init process flags.

Provided that you can answer questions above, makes sense, sure. But then su-exec doesn't need to be installed anymore.

File: slim/defaults/configs/znc.conf [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	PrependTimestamp = true
+	QuitMsg = %znc%
+	RealName = Real Name
+	StatusPrefix = *
+	TimestampFormat = [%H:%M:%S]
+
+	<Pass password>
+		Hash = cc6b37824eeda0b0ec028e18e98684ec71a3dbef57671f7c8d70af386455149e
+		Method = SHA256
+		Salt = hkXd849mvtdBhHgCQHIr
```
Comment (psychon) [inline slim/defaults/configs/znc.conf:59]: As far as I know, ZNC still supports unhashed passwords. That make sense to be use here. However, no idea what the state of that support is. It should only exist for backwards compatibility....

**Comment by ironcladlou** [comment]
@DarthGandalf
>>If the data volume doesn’t contain a znc.conf, populate the volume with a sane default configuration set.

>Ok. But if user passes --makeconf, it shouldn't copy this default config, and immediately complain that config already exists.

I learned two things from this statement:

1. The image can (should?) work as drop-in replacement of the `znc` binary in terms of how the user executes it
2. The `--makeconf` option exists and probably already does what I'm doing here manually

I need to experiment more with this: if I could run the image once with `--makeconf` to generate defaults, that would much more cleanly do what I was out to achieve.

>>Expose the container port by default.

>What difference does it make? Any port can be forwarded with -p, regardless of what ports are exposed.

Nothing major, just allows the use of `-P` to auto-forward based on what's declared via `EXPOSE`.

>> Remove chown/chmod commands from entrypoint (these break most non-host mounted volume scenarios).

> Break how? What are these scenarios?
> So what will happen if user is wrong?

The practical requirement seems to be that the `znc` user has RW permissions on the data directory. The data dir could be root owned but user writable: for example, a GlusterFS volume in Kuberentes. The `znc` process running as the `znc` user can work with the mounted data directory just fine with RW permissions, but can't change the ownership of the directory itself.

Is there some reason the data directory needs *ownership* that I'm not understanding?

>>Run as the ZNC user via USER directive rather than init process flags.

>Provided that you can answer questions above, makes sense, sure. But then su-exec doesn't need to be installed anymore.

Agree re: `su-exec`.

Thanks for the review and discussion. ZNC rocks!

**Comment by DarthGandalf** [comment]
> The image can (should?) work as drop-in replacement of the znc binary in terms of how the user executes it

I think so, yeah.

> The --makeconf option exists and probably already does what I'm doing here manually

Yes, that's how initial config is supposed to be created, not by hand. It's documented both at http://znc.in and at https://hub.docker.com/_/znc/

> Is there some reason the data directory needs ownership that I'm not understanding?

Not really. The chown is there to ensure znc user has RW access to the directory. It's not the case if previously ZNC was running without docker, and stores its settings in `~/.znc` which is chmodded to be owner-accessible only, as it contains SSL cert, passwords, etc. UID of znc user inside docker is probably different from UID of user who owned that dir before.
That said, chowning it to another UID of system user isn't great either, probably migration path needs to be different from `-v $HOME/.znc:/znc-data`.

> Run as the ZNC user via USER directive rather than init process flags.

One more thing: in full image, need to revert USER back to root before calling apk, and then set it to `znc` again.

*Edit: s/chmod/chown/ where appropriate*

**Comment by ironcladlou** [comment]
Will open a new PR when I have time to rethink the approach based on feedback. Thanks!

**Pull request closed without merging**